src/rust/ide/lib/span-tree/src/node/kind.rs

```diff
@@ -0,0 +1,303 @@
+    pub fn is_argument        (&self) -> bool { matches!(self,Self::Argument{..}) }
+    pub fn is_insertion_point (&self) -> bool { matches!(self,Self::InsertionPoint{..}) }
+
+    /// Match the value with `Kind::InsertionPoint{..}` but not
+    /// `Kind::InsertionPoint(ExpectedArgument(_))`.
+    pub fn is_positional_insertion_point(&self) -> bool {
+        self.is_insertion_point() && !self.is_expected_argument()
+    }
+
+    /// Match the value with `Kind::InsertionPoint(ExpectedArgument(_))`.
+    pub fn is_expected_argument(&self) -> bool {
+        match self {
+            Self::InsertionPoint(t) => t.kind.is_expected_argument(),
+            _                       => false
+        }
+    }
+}
+
+
+// === API ===
+
+impl Kind {
+    /// Smart name getter.
+    pub fn name(&self) -> Option<&String> {
+        match self {
+            Self::Argument       (t) => t.name.as_ref(),
+            Self::InsertionPoint (t) => t.name.as_ref(),
+            _                        => None,
+        }
+    }
+
+    /// Smart type getter.
+    pub fn tp(&self) -> Option<&String> {
+        match self {
+            Self::Argument       (t) => t.tp.as_ref(),
+            Self::InsertionPoint (t) => t.tp.as_ref(),
+            _                        => None,
+        }
+    }
+
+    /// Smart removable flag getter.
+    pub fn removable(&self) -> bool {
+        match self {
+            Self::This     (t) => t.removable,
+            Self::Argument (t) => t.removable,
+            _                  => false,
+        }
+    }
+
+    /// Smart `ArgumentInfo` getter. Returns `None` if the node could not be attached with the
+    /// information.
+    pub fn argument_info(&self) -> Option<ArgumentInfo> {
+        match self {
+            Self::Argument       (t) => Some(ArgumentInfo::new(t.name.clone(),t.tp.clone())),
+            Self::InsertionPoint (t) => Some(ArgumentInfo::new(t.name.clone(),t.tp.clone())),
+            _                        => None
+        }
+    }
+
+    /// Smart `ArgumentInfo` setter. Returns bool indicating whether the operation was possible
+    /// or was skipped.
+    pub fn set_argument_info(&mut self, argument_info:ArgumentInfo) -> bool {
+        match self {
+            Self::Argument(t) => {
+                t.name = argument_info.name;
+                t.tp   = argument_info.tp;
+                true
+            },
+            Self::InsertionPoint(t) => {
+                t.name = argument_info.name;
+                t.tp   = argument_info.tp;
+                true
+            },
+            _ => false
+        }
+    }
+}
+
+
+// === Impls ===
+
+impl Default for Kind {
+    fn default() -> Self {
+        Self::insertion_point().into()
+    }
+}
+
+
+
+// ============
+// === This ===
+// ============
+
+/// Kind representing "this" node. For example, in the following expressions, `foo` is considered
+/// "this": `foo.bar`, `foo + bar`, `foo.+ bar`.
```

Comment:
currently, in `bar foo` we also mark `foo` as This. Perhaps we should only when they are methods though, however that cannot be deduced from AST.